src/Stratis.Bitcoin.Features.BlockStore/BlockStoreCache.cs

```diff
@@ -34,18 +38,42 @@ public class BlockStoreCache : IBlockStoreCache
         /// <summary>Provider of time functions.</summary>
         protected readonly IDateTimeProvider dateTimeProvider;
 
-        public BlockStoreCache(IBlockRepository blockRepository, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
-            : this(blockRepository, new MemoryCache(new MemoryCacheOptions()), dateTimeProvider, loggerFactory)
-        {
-        }
+        /// <summary>The maximum amount of blocks the cache can contain.</summary>
+        public readonly int MaxCacheBlocksCount;
+
+        /// <summary>Entry options for adding blocks to the cache.</summary>
+        private readonly MemoryCacheEntryOptions blockEntryOptions;
 
-        public BlockStoreCache(IBlockRepository blockRepository, IMemoryCache memoryCache, IDateTimeProvider dateTimeProvider, ILoggerFactory loggerFactory)
+        /// <summary>Specifies amount to compact the cache by when the maximum size is exceeded.</summary>
+        /// <remarks>For example value of 0.8 will let cache remove 20% of all items when cache size is exceeded.</remarks>
+        private readonly double CompactionPercentage = 0.8;
+
+        public BlockStoreCache(
+            IBlockRepository blockRepository,
+            IDateTimeProvider dateTimeProvider,
+            ILoggerFactory loggerFactory,
+            NodeSettings nodeSettings,
+            IMemoryCache memoryCache = null)
         {
             Guard.NotNull(blockRepository, nameof(blockRepository));
-            Guard.NotNull(memoryCache, nameof(memoryCache));
 
-            this.blockRepository = blockRepository;
+            this.MaxCacheBlocksCount = nodeSettings.ConfigReader.GetOrDefault("maxCacheBlocksCount", 300);
+
             this.cache = memoryCache;
+            if (this.cache == null)
+            {
+                var memoryCacheOptions = new MemoryCacheOptions()
+                {
+                    SizeLimit = this.MaxCacheBlocksCount,
+                    CompactionPercentage = this.CompactionPercentage
+                };
+
+                this.cache = new MemoryCache(memoryCacheOptions);
+            }
+
+            this.blockEntryOptions = new MemoryCacheEntryOptions() {AbsoluteExpirationRelativeToNow = TimeSpan.FromMinutes(60), Size = 1 };
```

Comment:
more magic here
and space after `{`

Reply:
btw 60 minutes is 3-9 blocks in bitcoin, what will happen with that cache entry then?
will it be removed? if so, we need much higher number here
or what is this 60 minutes actually?

Reply:
>and space after {

Fixed.

>more magic here

Well, the 'AbsoluteExpirationRelativeToNow' is commented so it should be obvious what this number does imho.

Reply:
>btw 60 minutes is 3-9 blocks in bitcoin, what will happen with that cache entry then?
will it be removed? if so, we need much higher number here

Correct, it will be removed. 
We don't need a higher number because the chances of needing that block in 60 minutes are very small. Usually we only need to access the block to propagate it if the peer asks so even old value of 10 minutes sounded good enough for me.

Reply:
it has been somewher, probably in bitcoin mailling list, when discussing the pruned nodes, that it is reasonable to have something like a day of blocks because many peers are asking for them

this is slightly different use case, but we can use a similar reasoning - which was that the more recent the block, the higher chance for it to be needed/requested

so 3-9 blocks seems not that good actually, if peers connect to us asking for blocks because they are not always online and need sync, we can use this cache, so 60 minutes seems too low to me

of course for the use case of block propagation at the tip, last 3 blocks would be sufficient, but i'd be happier if this cache was reused for peer syncing as well

so this is used when we receive "getdata" message and for that i'd like to see something like 20-50 blocks, it makes sense imho

Reply:
Value of 24h sounds good?

Reply:
actually yes, 24 hours makes sense, in fact infinity would make sense too because our primary criterion is item count, not time, so time could be arbitrary except not low, so yes 24h is ok

Reply:
Fixed